include/simdjson/dom/parser-inl.h

```diff
@@ -168,17 +175,31 @@ inline bool parser::allocate_capacity(size_t capacity, size_t max_depth) noexcep
   return !allocate(capacity, max_depth);
 }
 #endif // SIMDJSON_DISABLE_DEPRECATED_API
-inline error_code parser::ensure_capacity(size_t desired_capacity) noexcept {
+
+inline error_code parser::ensure_capacity_with_document(size_t desired_capacity) noexcept {
+  return ensure_capacity(doc, desired_capacity);
+}
+
+
+inline error_code parser::ensure_capacity(document& target_document, size_t desired_capacity) noexcept {
+  // 1. It is wasteful to allocate a document and a parser for documents spanning less than MINIMAL_DOCUMENT_CAPACITY bytes.
+  // 2. If we allow desired_capacity = 0 then it is possible to exit this function with implementation == nullptr.
```

Comment:
I was treating allocate(0) as a way to "garbage collect," zeroing out the memory of a parser you weren't going to use for a long time or something. I don't even know if it matters, and there are certainly other ways of doing it.

MINIMAL_DOCUMENT_CAPACITY is a good idea! There's an interaction with the max capacity you can set in the document, if max_capacity is smaller ... but I don't think it's worth addressing, especially since MINIMAL_DOCUMENT_CAPACITY is intended to be ... well, minimal.

Reply:
Since we allocated memory dynamically, using as little memory as possible given the document encountered, then it is possible that we could be allocating ridiculously small amounts of memory.
I think we can validate the max too.

Reply:
I have added a commit that ensures that you can't set the max capacity to a value lower than MINIMAL_DOCUMENT_CAPACITY. So you can't accidentally set it to zero, for example, and get a dysfunctional parser.

Reply:
It would make sense to have a 'clear' function that flushes the memory.